src/main.cairo

```diff
@@ -1,32 +1,182 @@
+    return (res,);
 }
 
+//
+// Externals
+//
+
+// @notice Activate or deactivate renewals for a domain
+// @param domain Domain to activate or deactivate renewals for
+// @dev callable by owner of domain
+@external
+func toggle_renewals{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain: felt,
+) {
+    alloc_locals;
+    let (caller) = get_caller_address();
+    let (contract) = naming_contract.read();
+
+    let (prev_renew) = renews.read(caller, domain);
+    renews.write(caller, domain, 1 - prev_renew);
+
+    ToggledRenewal.emit(domain, caller, 1 - prev_renew);
+
+    return ();
+} 
+
+// @notice Renew a domain
+// @param domain Domain to renew
+// @dev callable by anyone, but will only renew if renewals are activated for domain
 @external
-func increase_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
-    amount: felt
+func renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    root_domain: felt, renewer: felt,
 ) {
-    with_attr error_message("Amount must be positive. Got: {amount}.") {
-        assert_nn(amount);
+    alloc_locals;
+    let (naming) = naming_contract.read(); 
+    let (can_renew) = renews.read(renewer, root_domain);
+
+    with_attr error_message("Renewer has not activated renewals for this domain") {
+        assert_not_zero(can_renew);
+    }
+
+    let (expiry) = Naming.domain_to_expiry(naming, 1, cast(new(root_domain), felt*));
+    let (block_timestamp) = get_block_timestamp();
+    with_attr error_message("Domain is not set to expire within a month") {
+        assert_le(expiry, block_timestamp + 86400 * 30);
+    }
+
+    let (pricing) = pricing_contract.read();
+    let (erc20, renewal_price) = Pricing.compute_renew_price(pricing, root_domain, 365);
+
+    let (contract) = get_contract_address();
+    with_attr error_message("Error transferring tokens from renewer to contract") {
+        IERC20.transferFrom(erc20, renewer, contract, renewal_price);
     }
 
-    let (res) = balance.read();
-    balance.write(res + amount);
+    Naming.renew(naming, root_domain, 365);
+    
+    DomainRenewed.emit(root_domain, renewer, 365);
+
     return ();
 }
 
-@view
-func get_balance{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (res: felt) {
-    let (res) = balance.read();
-    return (res,);
+@external
+func batch_renew{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    domain_len: felt, domain: felt*, renewer_len: felt, renewer: felt*
+) {
+    with_attr error_message("domain and renewer array must have same length") {
+        assert domain_len = renewer_len;
+    }
+    return _batch_renew_iter(domain_len, domain, renewer_len, renewer);
 }
 
-@constructor
-func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() {
-    balance.write(0);
+func _batch_renew_iter{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
```

Comment:
We need this one to send one tx each day on the bot right ?

Reply:
Yes